[mlir][SCF] Do not peel already peeled loops

matthias-springer opened the pull request:
Loop peeling is not beneficial if the step size already divides "ub - lb". There are currently some simple checks to prevent peeling in such cases when lb, ub, step are constants. This commit adds support for IR that is the result of loop peeling in the general case; i.e., lb, ub, step do not necessarily have to be constants.

This change adds a new affine_map simplification rule for semi-affine maps that appear during loop peeling and are guaranteed to evaluate to a constant zero. Affine maps such as:
```
(1) affine_map<()[ub, step] -> ((ub - ub mod step) mod step)
(2) affine_map<()[ub, lb, step] -> ((ub - (ub - lb) mod step - lb) mod step)
(3)                                         ^ may contain additional summands
```
Other affine maps with modulo expressions are not supported by the new simplification rule.

This fixes #71469.

matthias-springer commented:
Note: This new simplification rule is based on the tree-based representation of affine maps. I was trying to implement the simplification logic based on the flattened representation (with the hope that more simplifications around modulo expressions could be supported), but I think that is not possible if the RHS of the modulo expression is not a constant.

Please let me know if you know a better way to do this.

Groverkss commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -444,12 +444,88 @@ static AffineExpr symbolicDivide(AffineExpr expr, unsigned symbolPos,
+/// `lhs` is divisible by `rhs`. This is a common pattern in the resulting IR
+/// after loop peeling.
+///
+/// Example: lhs = ub - ub % step
```
Is this supposed to be `ub - lb`?

matthias-springer replied:
This example is for the frequent case where `lb = 0`. (`scf.for` often have a lb of zero.) This function is not specific to loop peeling, so you can think of `ub` and `step` as just `s0` and `s1`.

Groverkss replied:
I see. Here it makes sense.

Groverkss commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -444,12 +444,88 @@ static AffineExpr symbolicDivide(AffineExpr expr, unsigned symbolPos,
+///
+/// Example: lhs = ub - ub % step
+///          rhs = step
+///       => (ub - ub % step) % step is guaranteed to evaluate to 0.
```
`ub - lb`?

matthias-springer replied:
Same as above, think of this as `(s0 - s0 mod s1) mod s1`. This affine expr appears during loop peeling of `scf.for` loops with `lb = 0`.

Groverkss replied:
Makes sense.

Groverkss commented on mlir/lib/IR/AffineExpr.cpp:
```diff
@@ -444,12 +444,88 @@ static AffineExpr symbolicDivide(AffineExpr expr, unsigned symbolPos,
+/// Example: lhs = ub - ub % step
+///          rhs = step
+///       => (ub - ub % step) % step is guaranteed to evaluate to 0.
+static bool isModOfModSubtraction(AffineExpr lhs, AffineExpr rhs,
```
This looks very similar to computing a bound over lhs. Do you think we can improve https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/AffineExpr.cpp#L1441 to do something like this? I think it handles bounds already. Maybe we can teach it to return a symbolic result instead?

matthias-springer replied:
That is tricky because we need to be able to handle cases such as:
```
Mod(Add(Add(a, Mul(-1, Mod(Add(a, b), c))), b), c)
```

Because of the nested `Mod` we cannot use the flattened representation to handle `Mul` and `Add`.

Groverkss replied:
Right, that function also does some non-flattened floordiv/mod handelling as well. I was hoping we could combine both of these functions to get better bounds.

I would appreciate A TODO for merging these functions for now, but that is optional.

Groverkss reviewed (approved):
LGTM

matthias-springer committed:
[mlir][SCF] Do not peel already peeled loops

Loop peeling is not beneficial if the step size already divides "ub - lb". There are currently some simple checks to prevent peeling in such cases when lb, ub, step are constants. This commit adds support for IR that is the result of loop peeling in the general case; i.e., lb, ub, step do not necessarily have to be constants.

This change adds a new affine_map simplification rule for semi-affine maps that appear during loop peeling and are guaranteed to evaluate to a constant zero. Affine maps such as `(ub - ub mod step) mod step`.

dcaballe commented:
Thanks! @banach-space is verifying that this works

banach-space commented:
> Thanks! @banach-space is verifying that this works

I actually haven't been able to trigger "double" peeling in IREE, but have verified this with a small standalone repro with scalable sizes:
```mlir
func.func @matmul(%A: tensor<?x?xf32>, %B: tensor<?x?xf32>, %C: tensor<?x?xf32>) -> tensor<?x?xf32> {
  %out = linalg.matmul ins(%A, %B: tensor<?x?xf32>, tensor<?x?xf32>)
            outs(%C: tensor<?x?xf32>) -> tensor<?x?xf32>
  return %out : tensor<?x?xf32>
}

module attributes {transform.with_named_sequence} {
  transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
    %matmul_0 = transform.structured.match ops{["linalg.matmul"]} in %arg1 : (!transform.any_op) -> !transform.any_op
    %1, %loop_1, %loop_2, %loop_3 = transform.structured.tile_using_for %matmul_0 [8, [32], 1] : (!transform.any_op)
      -> (!transform.any_op, !transform.op<"scf.for">, !transform.op<"scf.for">,!transform.op<"scf.for">)
    %main_loop, %remainder_loop = transform.loop.peel %loop_2 : (!transform.op<"scf.for">) -> (!transform.op<"scf.for">, !transform.op<"scf.for">)
    // "Double" peeling - fails after this change, which is correct.
    %main_loop_2, %remainder_loop_2 = transform.loop.peel %main_loop : (!transform.op<"scf.for">) -> (!transform.op<"scf.for">, !transform.op<"scf.for">)

    transform.yield
  }
}
```

 Thanks for the fix @matthias-springer !
